### PR TITLE
[MRESOLVER-354] Document Expected Checksums in Resolver

### PR DESCRIPTION
**Pull request by cstamas**
Add documentation to new resolver transport features like "provided" and "remote included" checksums. Also explain "trusted" checksums.

---

https://issues.apache.org/jira/browse/MRESOLVER-354

### CODE REVIEW

File: src/site/markdown/expected-checksums.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+then obtains somehow the "expected" checksum (for same payload)
+and compares the two.
+
+This page covers all the "expected" checksum varieties.
+
+
+## Transport Checksum Strategies
+
+Historically, the "obtain expected checksum" was implemented as simple HTTP GET 
+request against Artifact checksum URL (Artifact URL appended by ".sha1"). This logic
```
Comment (michael-o) [inline src/site/markdown/expected-checksums.md:38]: Note that this is actually transport specific.

File: src/site/markdown/expected-checksums.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+is still present in current Resolver, but is "decorated" and extended in multiple 
+ways.
+
+Resolver has broadened the "obtain" step for "expected" checksum with two new strategies,
+so the three expected checksum kinds in transport are: "Provided", "Remote Included" and 
+"Remote External". All these strategies provide the source of "expected" checksum, 
+but it differs **how** Resolver obtains these.
+
+The new **Provided** kind of expected checksums are "provided" to resolver by some alternative
+means, possibly ahead of any transport operation. There is an SPI interfacce that users may
```
Comment (hazendaz) [inline src/site/markdown/expected-checksums.md:48]: Spelling on 'interfacce' should be 'interface'

File: src/site/markdown/expected-checksums.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+Trusted checksums is a SPI component that is able to deliver "expected" checksums 
+for given Artifact, without use of any transport API element. In other words, this
+API is not bound to transport.
+
+Since they almost on-to-one into transport "Provided Checksum" strategy, resolver provides 
+implementation that simply delegates Provided to Trusted checksums (is able to
+make Provided and Trusted checksums equivalent, as transport is regarded).
+
+But the biggest game changer of Trusted Checksumms is their transport independence, that they
```
Comment (hazendaz) [inline src/site/markdown/expected-checksums.md:116]: Spelling with 'Checksumms' should be 'Checksums'

File: src/site/markdown/expected-checksums.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+can be utilized in places where there is no transport happening at all.
+
+One of such uses of Trusted Checksums in Resolver is in ArtifactResolver "post processing".
+This new functionality, at the cost of checksumming overhead, is able to validate all
+the resolved artifacts againstr Truated Checksums, thus, making sure that all resolved
```
Comment (hazendaz) [inline src/site/markdown/expected-checksums.md:121]: Few spelling issues in line 121.

File: src/site/markdown/expected-checksums.md (hunk 1 above)
Comment (cstamas) [reply src/site/markdown/expected-checksums.md:38]: removed mention of HTTP GET

File: src/site/markdown/expected-checksums.md (hunk 2 above)
Comment (cstamas) [reply src/site/markdown/expected-checksums.md]: tx, fixed

File: src/site/markdown/expected-checksums.md (hunk 4 above)
Comment (cstamas) [reply src/site/markdown/expected-checksums.md]: tx, fixed

File: src/site/markdown/expected-checksums.md (hunk 3 above)
Comment (cstamas) [reply src/site/markdown/expected-checksums.md]: tx, fixed

File: src/site/markdown/expected-checksums.md [hunk 5]
```diff
@@ -45,7 +45,7 @@ so the three expected checksum kinds in transport are: "Provided", "Remote Inclu
 but it differs **how** Resolver obtains these.
 
 The new **Provided** kind of expected checksums are "provided" to resolver by some alternative
-means, possibly ahead of any transport operation. There is an SPI interfacce that users may 
+means, possibly ahead of any transport operation. There is an SPI interface that users may
```
Comment (michael-o) [inline src/site/markdown/expected-checksums.md:53]: the word interface is redudant since SPI ends with interface already

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/site/markdown/expected-checksums.md (hunk 5 above)
Comment (cstamas) [reply src/site/markdown/expected-checksums.md]: changed, "SPI interface" -> "SPI extension point"

Comment (michael-o) [reply src/site/markdown/expected-checksums.md]: Good!

File: src/site/markdown/expected-checksums.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+ways.
+
+Resolver has broadened the "obtain checksum" step for "expected" checksum with two new strategies,
+so the three expected checksum kinds in transport are: "Provided", "Remote Included" and 
+"Remote External". All these strategies provide the source of "expected" checksum, 
+but it differs **how** Resolver obtains these.
+
+The new **Provided** kind of expected checksums are provided to resolver by some alternative
+means, possibly ahead of any transport operation. There is an SPI extension point that users may 
+implement, to have own ways to provide checksums to resolver. Alternatively, one may use Resolver out of the
```
Comment (slawekjaranowski) [inline src/site/markdown/expected-checksums.md:54]: An example implementation or demo  - how to implement will be helpful, can be on separated document.

File: src/site/markdown/expected-checksums.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+But the biggest game changer of Trusted Checksums is their transport independence, that they
+can be utilized in places where there is no transport happening at all.  One of such uses of 
+Trusted Checksums is ArtifactResolver post-processing.
+This new functionality, at the cost of checksum calculation overhead, is able to validate all
+the resolved artifacts against Trusted Checksums, thus, making sure that all resolved
+artifacts are "validated" with some known (possibly even cryptographically strong) checksum
+provided by user.  This new feature may become handy in cases when user does not trust the local
+repository, as it may be shared with some other unknown or even untrusted party.
+
+The Trusted Checksums provide two source implementations out of the box.
```
Comment (slawekjaranowski) [inline src/site/markdown/expected-checksums.md:130]: example how to configure, which property to set ... etc will be appreciated, can be as link to other document

**Review by slawekjaranowski — APPROVED** [review]
Generally ok, it describes used mechanizm and possibility from theoretically side. 
Examples how to use and configure will be appreciated. 
Of course on separate documents for each case as next step.

**Comment by cstamas** [comment]
@slawekjaranowski re examples, all the new features (RRF, TrustedChecksums) are kinda incomplete without https://issues.apache.org/jira/browse/MRESOLVER-304 (or https://issues.apache.org/jira/browse/MNG-6303) as ultimate goal would be to commit checksums, filter config along with project sources, that is impossible today.

So I'd rather wait with examples. But still, this doco is useful, as it explains at least one confusion users hit (and reported as "bug" while it was actually OK) https://lists.apache.org/thread/nsbormhpfgqq19qzgo4rx12lg28qv0tb

**Pull request merged**